x/protocolpool/keeper/keeper.go

```diff
@@ -67,3 +88,110 @@ func (k Keeper) GetCommunityPool(ctx context.Context) (sdk.Coins, error) {
 	}
 	return k.bankKeeper.GetAllBalances(ctx, moduleAccount.GetAddress()), nil
 }
+
+func (k Keeper) claimFunds(ctx context.Context, recipient sdk.AccAddress) (amount sdk.Coin, err error) {
+	// get claimable funds from distribution info
+	amount, err = k.getClaimableFunds(ctx, recipient)
+	if err != nil {
+		return sdk.Coin{}, err
+	}
+
+	// distribute amount from feepool
+	err = k.DistributeFromFeePool(ctx, sdk.NewCoins(amount), recipient)
+	if err != nil {
+		return sdk.Coin{}, err
+	}
+
+	return amount, nil
+}
+
+func (k Keeper) getClaimableFunds(ctx context.Context, recipient sdk.AccAddress) (amount sdk.Coin, err error) {
+	sdkCtx := sdk.UnwrapSDKContext(ctx)
+
+	budget, err := k.BudgetProposal.Get(ctx, recipient)
+	if err != nil {
+		if errors.Is(err, collections.ErrNotFound) {
+			return sdk.Coin{}, fmt.Errorf("no budget found for recipient: %s", recipient.String())
+		}
+		return sdk.Coin{}, err
+	}
+
+	// check if the distribution is completed
+	if budget.Tranches <= 0 {
+		// remove the entry of budget ended recipient
+		err := k.BudgetProposal.Remove(ctx, recipient)
+		if err != nil {
+			return sdk.Coin{}, err
+		}
+		// Return the end of the budget
+		return sdk.Coin{}, fmt.Errorf("budget ended for recipient: %s", recipient.String())
+	}
+
+	currentTime := sdkCtx.BlockTime().Unix()
+	startTime := budget.StartTime
+	// Check if the start time is reached
+	if uint64(currentTime) < startTime {
+		return sdk.Coin{}, fmt.Errorf("distribution has not started yet")
+	}
+
+	// Calculate the time elapsed since the start time
+	timeElapsed := uint64(currentTime) - startTime
+	// Check the time elapsed has passed period length
+	if timeElapsed >= budget.Period {
+		// Calculate how many periods have passed
+		periodsPassed := timeElapsed / budget.Period
+
+		if periodsPassed > 0 {
+			// Calculate the amount to distribute for all passed periods
+			coinsToDistribute := math.NewInt(int64(periodsPassed)).Mul(budget.TotalBudget.Amount.QuoRaw(int64(budget.Tranches)))
+			amount := sdk.NewCoin(budget.TotalBudget.Denom, coinsToDistribute)
+
+			// update the budget's remaining tranches
+			budget.Tranches -= periodsPassed
+
+			// update the TotalBudget amount
+			budget.TotalBudget.Amount = budget.TotalBudget.Amount.Sub(coinsToDistribute)
+
+			// Update the start time for the budget
+			budget.StartTime += budget.Period
```

Comment:
Can we have a more descriptive comment?
I understand why we update the start time, because how are calculated the periods, but it took me a pause to get it, as it sounds weird to update the start time.
Otherwise maybe a budget should have another field, called LastClaim, or something to be different from the start time, and use that value once the startTime has passed. This would make the code more readable and logical imho.

Reply:
I understand that it seems odd to update the start time, how about we change this field to `ClaimTime` with more description?
Otherwise, with adding one more field `LastClaim` we introduce additional functionality that claim can be started at start time or it can be after a defined delay. And we should  be checking multiple conditions(i.e., LastClaim>=StartTime, time elapsed since start time has passed period length, time elapsed since last claim time has also passed period length, etc.) making it more complex.

Reply:
I don't think it would make it more complex. The first claim will check startTime indeed, but after the first claim, it will anyway always be true. The second check will only check against LastClaim if set (so from second claim and on).

If you modify startTime to ClaimTime that would be off by one, as the StartTime isn't the ClaimTime for the first period (you can claim after the startTime)